### PR TITLE
Openshift install uninstall

### PR DESCRIPTION
**Pull request by myakove**
Install and uninstall IPI Openshift cluster on AWS, support multiple clusters.

### CODE REVIEW

File: app/cli.py [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+import yaml
+from click_dict_type import DictParamType
+from clouds.aws.aws_utils import set_and_verify_aws_credentials
+from ocp_utilities.utils import run_command
+
+
+def generate_cluster_dir_path(clusters, base_directory):
+    for _cluster in clusters:
+        cluster_name = _cluster["name"]
+        platform = _cluster["platform"]
```
Comment (rnetser) [inline app/cli.py:18]: can be removed

File: app/cli.py [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+    if not (install or uninstall):
+        raise ValueError("One of install/uninstall must be specified")
+
+    if debug:
+        set_debug_os_flags()
+
+    base_install_config = update_base_install_config(pull_secret_file=pull_secret_file)
+    clusters = generate_cluster_dir_path(
+        clusters=cluster, base_directory=clusters_install_data_directory
+    )
```
Comment (rnetser) [inline app/cli.py:206]: for uninstall we assume we have the proper file in place; we can call `update_base_install_config` from `create_install_config_file`?

File: app/cli.py [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+
+
+def download_openshift_install_binary(clusters, pull_secret_file):
+    versions = set()
+    for cluster in clusters:
+        versions.add(cluster["version"])
+
+    for version in versions:
+        binary_dir = os.path.join("/tmp", version)
+        openshift_install_str = "openshift-install"
```
Comment (rnetser) [inline app/cli.py:52]: can be outside the loop

File: app/cli.py [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+
+
+def uninstall_openshift(cluster_data):
+    directory = cluster_data["install-dir"]
+    binary_path = cluster_data["openshift-install-binary"]
+    return run_command(
+        command=shlex.split(f"{binary_path} destroy cluster --dir {directory}"),
+        capture_output=False,
+    )[0]
+
```
Comment (rnetser) [inline app/cli.py:87]: dup code; we can have one fucntion which does both

File: app/cli.py [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+
+def get_base_install_config_data():
+    with open("app/manifests/install-config.yaml") as fd:
+        return yaml.safe_load(fd)
+
+
+def set_debug_os_flags():
+    os.environ["OCM_PYTHON_WRAPPER_LOG_LEVEL"] = "DEBUG"
+    os.environ["OPENSHIFT_PYTHON_WRAPPER_LOG_LEVEL"] = "DEBUG"
```
Comment (rnetser) [inline app/cli.py:118]: for now we do not use these tools; we can remove

File: app/cli.py [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+    is_flag=True,
+    show_default=True,
+)
+@click.option(
+    "--clusters-install-data-directory",
+    help="""
+\b
+Path to cluster install data
+    For install this will be used to store the install data.
+    For uninstall this will be used to uninstall the cluster and must be provided.
```
Comment (rnetser) [inline app/cli.py:142]: `must be provided.` we have default for this so can be removed.

File: app/cli.py [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+Format to pass is:
+    'name=cluster1;baseDomain=aws.domain.com;platform=aws;region=us-east-2;version=4.14.0-ec.2'
+Required parameters:
+    name: Cluster name.
+    baseDomain: Base domain for the cluster.
+    platform: Cloud platform to install the cluster on. (Currently only AWS supported).
+    region: Region to use for the cloud platform.
+    version: Openshift cluster version to install
+
+Every parameter in install-config.yaml can be override by the user.
```
Comment (rnetser) [inline app/cli.py:175]: Nested parameters should have "_" as a path delimiter

File: app/manifests/install-config.yaml [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+  - cidr: 10.128.0.0/14
+    hostPrefix: 23
+  machineNetwork:
+  - cidr: 10.0.0.0/16
+  networkType: OVNKubernetes
+  serviceNetwork:
+  - 172.30.0.0/16
+platform:
+#  aws:
+#    region: us-west-1
```
Comment (rnetser) [inline app/manifests/install-config.yaml:29]: can we remove this?

File: pyproject.toml [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+repository = "https://github.com/RedHatQE/openshift-cli-installer"
+
+[tool.poetry.scripts]
+cli = "app.cli:main"
+
+[tool.poetry.dependencies]
+python = "^3.8"
+colorlog = "*"
+openshift-python-wrapper = "*"
+openshift-python-utilities = "*"
```
Comment (rnetser) [inline pyproject.toml:34]: can be removed

**Review by rnetser — CHANGES_REQUESTED** [review]

File: app/cli.py (hunk 3 above)
Comment (myakove) [reply app/cli.py:52]: How? we create folder for each version

File: app/cli.py (hunk 4 above)
Comment (myakove) [reply app/cli.py:87]: done

File: app/cli.py (hunk 7 above)
Comment (myakove) [reply app/cli.py:175]: Done

File: pyproject.toml (hunk 9 above)
Comment (myakove) [reply pyproject.toml:34]: Done

File: app/cli.py (hunk 1 above)
Comment (myakove) [reply app/cli.py:18]: Done

File: app/cli.py (hunk 2 above)
Comment (myakove) [reply app/cli.py:206]: For uninstall we use the first call of `generate_cluster_dir_path`, I don't understand what you mean here

File: app/cli.py (hunk 3 above)
Comment (omrirh) [reply app/cli.py:52]: `openshift_install_str` var is constant to the loop

File: app/cli.py [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+    )[0]
+
+
+def create_install_config_file(clusters, base_install_config):
+    for _cluster in clusters:
+        base_install_config_copy = copy.deepcopy(base_install_config)
+        cluster_name = _cluster["name"]
+        install_dir = _cluster["install-dir"]
+        base_install_config_copy["metadata"]["name"] = cluster_name
+        base_install_config_copy["baseDomain"] = _cluster["baseDomain"]
```
Comment (omrirh) [inline app/cli.py:95]: `"baseDomain"` can be re-used?

File: app/cli.py [hunk 11]
```diff
@@ -0,0 +1,239 @@
...
+        base_install_config_copy["platform"] = {
+            _cluster["platform"]: {"region": _cluster["region"]}
```
Comment (omrirh) [inline app/cli.py:97]: `"platform"` and `"region"` can be re-used

File: app/cli.py [hunk 12]
```diff
@@ -0,0 +1,239 @@
...
+    return run_command(
+        command=shlex.split(f"{binary_path} destroy cluster --dir {directory}"),
+        capture_output=False,
+    )[0]
+
+
+def create_install_config_file(clusters, base_install_config):
+    for _cluster in clusters:
+        base_install_config_copy = copy.deepcopy(base_install_config)
+        cluster_name = _cluster["name"]
```
Comment (omrirh) [inline app/cli.py:92]: `"name"` can be re-used

File: app/cli.py [hunk 13]
```diff
@@ -0,0 +1,239 @@
...
+    name: Cluster name.
+    baseDomain: Base domain for the cluster.
+    platform: Cloud platform to install the cluster on. (Currently only AWS supported).
+    region: Region to use for the cloud platform.
+    version: Openshift cluster version to install
+
+Every parameter in install-config.yaml can be override by the user.
+For example:
+    fips=true
+    compute_platform_aws_type=m5.xlarge
```
Comment (omrirh) [inline app/cli.py:178]: maybe we can add example for how to set number of machines?

File: app/cli.py (hunk 12 above)
Comment (myakove) [reply app/cli.py:92]: Prefer not to, save dict key as const is not add any value

File: app/cli.py (hunk 13 above)
Comment (myakove) [reply app/cli.py:178]: Done

**Review by rnetser — DISMISSED** [review]

File: Dockerfile [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+
+WORKDIR /openshift-cli-installer
+RUN mkdir clusters-install-data
+
+RUN python3 -m pip install pip --upgrade \
+    && curl -sSL https://install.python-poetry.org | python3 - \
+    && poetry config cache-dir /openshift-cli-installer \
+    && poetry config virtualenvs.in-project true \
+    && poetry config installer.max-workers 10 \
+    && poetry install
```
Comment (oharan2) [inline Dockerfile:33]: Just to understand - why not use `&& poetry export --without-hashes -n` here?

File: app/manifests/install-config-template.j2 [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+  - cidr: 10.128.0.0/14
+    hostPrefix: 23
+  machineNetwork:
+  - cidr: 10.0.0.0/16
+  networkType: OVNKubernetes
+  serviceNetwork:
+  - 172.30.0.0/16
+platform:
+  {{ platform }}:
+    region: {{ region }}
```
Comment (oharan2) [inline app/manifests/install-config-template.j2:34]: Could we set a default region?

File: app/cli.py [hunk 16]
```diff
@@ -0,0 +1,235 @@
...
+    pull_secret_file,
+    parallel,
+    cluster,
+    clusters_install_data_directory,
+):
+    """
+    Install/Uninstall Openshift cluster/s
+    """
+    set_and_verify_aws_credentials()
+    if not (install or uninstall):
```
Comment (oharan2) [inline app/cli.py:196]: Do we want to disable both being set? the code 'prioritize' install over uninstall if both are set to True (as flags)

Comment (oharan2) [reply app/cli.py:196]: Or make co existence of install&uninstall processes in the code

File: app/cli.py [hunk 17]
```diff
@@ -0,0 +1,235 @@
...
+    create_or_destroy_cluster(cluster_data=cluster_data, action="create")
+
+
+def uninstall_openshift(cluster_data):
+    create_or_destroy_cluster(cluster_data=cluster_data, action="destroy")
+
+
+def create_install_config_file(clusters, pull_secret_file):
+    ssh_key = get_local_ssh_key()
+    pull_secret = json.dumps(get_pull_secret_data(pull_secret_file=pull_secret_file))
```
Comment (oharan2) [inline app/cli.py:101]: As I understand the cli is getting a pull secret file indication and not reffering it to a specific cluster - is it "forcing" all cluster to be installed with the specified pull secret?

File: Dockerfile (hunk 14 above)
Comment (myakove) [reply Dockerfile:33]: Not adding any extra value

File: app/cli.py (hunk 17 above)
Comment (myakove) [reply app/cli.py:135]: the pull secret is to get the images from redhat registry

File: app/cli.py (hunk 16 above)
Comment (myakove) [reply app/cli.py:242]: right, fixed.

File: app/manifests/install-config-template.j2 (hunk 15 above)
Comment (myakove) [reply app/manifests/install-config-template.j2:34]: No, the user should set it

**Review by oharan2 — APPROVED** [review]

File: app/click_dict_type.py [hunk 18]
```diff
@@ -0,0 +1,79 @@
+# Taken from https://gist.github.com/welel/9cf860dd3f4d3e09f9b4305878b3a04e
+
+import click
+
+
+class DictParamType(click.ParamType):
```
Comment (omrirh) [inline app/click_dict_type.py:6]: should we reuse this class in `ocp-addons-operators-cli` ?

**Review by omrirh — APPROVED** [review]

**Review by rnetser — APPROVED** [review]

**Pull request merged**